File: core/src/main/java/com/github/fburato/highwheelmodules/core/analysis/StrictAnalyser.java
package com.github.fburato.highwheelmodules.core.analysis;

import com.github.fburato.highwheelmodules.core.algorithms.ModuleGraphTransitiveClosure;
import com.github.fburato.highwheelmodules.model.modules.Definition;
import com.github.fburato.highwheelmodules.model.modules.HWModule;
import com.github.fburato.highwheelmodules.model.modules.ModuleGraph;
import com.github.fburato.highwheelmodules.model.modules.TrackingModuleDependency;
import com.github.fburato.highwheelmodules.model.rules.NoStrictDependency;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import static com.github.fburato.highwheelmodules.core.analysis.AnalysisUtils.*;

public class StrictAnalyser implements Analyser {

    @Override
    public AnalyserModel.AnalysisResult analyse(AnalysisState state) {
        return analyseStrict(state);
    }

    private AnalyserModel.AnalysisResult analyseStrict(AnalysisState analysisState) {
        final ModuleGraphTransitiveClosure specTransitiveClosure = new ModuleGraphTransitiveClosure(
                analysisState.specGraph, append(analysisState.modules, analysisState.other));
        final ModuleGraphTransitiveClosure actualTransitiveClosure = new ModuleGraphTransitiveClosure(
                analysisState.actualGraph, append(analysisState.modules, analysisState.other));
        final List<AnalyserModel.EvidenceBackedViolation> dependencyViolations = getDependenciesViolations(
                specTransitiveClosure.diffPath(actualTransitiveClosure).get(), analysisState.other,
                analysisState.actualTrackingGraph);
        final List<AnalyserModel.ModuleConnectionViolation> noStrictDependencyViolations = getNoStrictDependencyViolations(
                actualTransitiveClosure, analysisState.noStrictDependencies, analysisState.other);
        final List<AnalyserModel.Metrics> metrics = getMetrics(analysisState.actualGraph, analysisState.modules,
                analysisState.actualGraph, analysisState.other);

        return new AnalyserModel.AnalysisResult(dependencyViolations, noStrictDependencyViolations, metrics);
    }

    private List<AnalyserModel.EvidenceBackedViolation> getDependenciesViolations(
            List<ModuleGraphTransitiveClosure.PathDifference> differences, HWModule other,
            ModuleGraph<TrackingModuleDependency> trackingGraph) {
        final List<AnalyserModel.EvidenceBackedViolation> evidenceBackedViolations = new ArrayList<>(
                differences.size());
        for (ModuleGraphTransitiveClosure.PathDifference difference : differences) {
            if (!difference.source.equals(other) && !difference.dest.equals(other)) {
                evidenceBackedViolations.add(new AnalyserModel.EvidenceBackedViolation(difference.source.name,
                        difference.dest.name, getNames(difference.firstPath), getNames(difference.secondPath),
                        getEvidence(trackingGraph, difference.source, difference.secondPath)));
            }
        }
        return evidenceBackedViolations;
    }

    private List<AnalyserModel.ModuleConnectionViolation> getNoStrictDependencyViolations(
            ModuleGraphTransitiveClosure transitiveClosure, Collection<NoStrictDependency> rules, HWModule other) {
        final List<AnalyserModel.ModuleConnectionViolation> noStrictDependencyViolations = new ArrayList<>();
        for (NoStrictDependency rule : rules) {
            if (!rule.source.equals(other) && !rule.dest.equals(other)
                    && transitiveClosure.minimumDistance(rule.source, rule.dest).get() == 1) {
                noStrictDependencyViolations
                        .add(new AnalyserModel.ModuleConnectionViolation(rule.source.name, rule.dest.name));
            }
        }
        return noStrictDependencyViolations;
    }
}

File: core/src/main/scala/com/github/fburato/highwheelmodules/core/analysis/StrictAnalyser.scala
package com.github.fburato.highwheelmodules.core.analysis

import com.github.fburato.highwheelmodules.core.algorithms.ModuleGraphTransitiveClosure
import com.github.fburato.highwheelmodules.core.algorithms.ModuleGraphTransitiveClosure.PathDifference
import com.github.fburato.highwheelmodules.model.modules.{HWModule, ModuleGraph, TrackingModuleDependency}
import com.github.fburato.highwheelmodules.model.rules.NoStrictDependency

import scala.jdk.CollectionConverters._

private[analysis] object StrictAnalyser extends Analyser {
  override def analyse(state: AnalysisState): AnalysisResult = {
    val specTransitiveClosure = new ModuleGraphTransitiveClosure(state.specGraph, (state.modules ++ Seq(state.other)).asJava)
    val actualTransitiveClosure = new ModuleGraphTransitiveClosure(state.actualGraph, (state.modules ++ Seq(state.other)).asJava)
    val dependencyViolations = getDependenciesViolations(specTransitiveClosure.diffPath(actualTransitiveClosure).get().asScala.toSeq, state.other, state.actualTrackingGraph)
    val noStrictDependencyViolations = getNoStrictDependencyViolation(actualTransitiveClosure, state.noStrictDependencies, state.other)
    val metrics = getMetrics(state.actualGraph, state.modules, state.actualGraph, state.other)
    AnalysisResult(dependencyViolations, noStrictDependencyViolations, metrics)
  }

  private def getDependenciesViolations(differences: Seq[PathDifference],
                                        other: HWModule,
                                        trackingGraph: ModuleGraph[TrackingModuleDependency]): Seq[EvidenceBackedViolation] = {
    differences
      .filter(d => d.source != other && d.dest != other)
      .map(d => EvidenceBackedViolation(d.source.name, d.dest.name,
        getNames(d.firstPath.asScala.toSeq),
        getNames(d.secondPath.asScala.toSeq),
        getEvidence(trackingGraph, d.source, d.secondPath.asScala.toSeq)))
  }

  private def getNoStrictDependencyViolation(transitiveClosure: ModuleGraphTransitiveClosure, rules: Seq[NoStrictDependency], other: HWModule): Seq[ModuleConnectionViolation] = {
    rules
      .filter(r => r.source != other && r.dest != other && transitiveClosure.minimumDistance(r.source, r.dest).get() == 1)
      .map(r => ModuleConnectionViolation(r.source.name, r.dest.name))
  }
}
